nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,344 @@
+  transformed_plan_pub_->publish(transformed_plan);
+
+  // Get the particular point on the path at the motion target distance and publish it
+  auto motion_target = getMotionTarget(params_->motion_target_dist, transformed_plan);
+  auto motion_target_point = nav2_graceful_motion_controller::createMotionTargetMsg(motion_target);
+  motion_target_pub_->publish(motion_target_point);
+
+  // Publish marker for slowdown radius around motion target for debugging / visualization
+  auto slowdown_marker = nav2_graceful_motion_controller::createSlowdownMarker(
+    motion_target,
+    params_->slowdown_radius);
+  slowdown_pub_->publish(slowdown_marker);
+
+  // Compute distance to goal
+  geometry_msgs::msg::PoseStamped goal_pose = transformed_plan.poses.back();
+  double dist_to_goal = std::hypot(goal_pose.pose.position.x, goal_pose.pose.position.y);
+
+  // If the distance to the goal is less than the motion target distance, i.e.
+  // the 'motion target' is the goal, then we skip the motion target orientation by pointing
+  // it in the same orientation that the last segment of the path
+  double angle_to_target = atan2(motion_target.pose.position.y, motion_target.pose.position.x);
+  if (params_->final_rotation && dist_to_goal < params_->motion_target_dist) {
+    geometry_msgs::msg::PoseStamped stl_pose =
+      transformed_plan.poses[transformed_plan.poses.size() - 2];
+    double dx = goal_pose.pose.position.x - stl_pose.pose.position.x;
+    double dy = goal_pose.pose.position.y - stl_pose.pose.position.y;
+    double yaw = std::atan2(dy, dx);
+    motion_target.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(yaw);
+  }
+
+  // Flip the orientation of the motion target if the robot is moving backwards
+  bool reversing = false;
+  if (params_->allow_backward && motion_target.pose.position.x < 0.0) {
+    reversing = true;
+    motion_target.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(
+      tf2::getYaw(motion_target.pose.orientation) + M_PI);
+  }
+
+  // Compute velocity command:
+  // 1. Check if we are close enough to the goal to do a final rotation in place
+  // 2. Check if we must do a rotation in place before moving
+  // 3. Calculate the new velocity command using the smooth control law
+  geometry_msgs::msg::TwistStamped cmd_vel;
+  cmd_vel.header = pose.header;
+  if (params_->final_rotation && (dist_to_goal < goal_dist_tolerance_ || goal_reached_)) {
+    goal_reached_ = true;
+    double angle_to_goal = tf2::getYaw(transformed_plan.poses.back().pose.orientation);
+    cmd_vel.twist = rotateToTarget(angle_to_goal);
+  } else if (params_->initial_rotation && // NOLINT
+    fabs(angle_to_target) > params_->initial_rotation_min_angle)
+  {
+    cmd_vel.twist = rotateToTarget(angle_to_target);
+  } else {
+    cmd_vel.twist = control_law_->calculateRegularVelocity(motion_target.pose, reversing);
+  }
+
+  // Transform local frame to global frame to use in collision checking
+  try {
```

Comment:
I don't understand this. What's up with this costmap transform? If we have some point `N` in the path, the path handler should have transformed it (from `map`) into the costmap's global frame (i.e. `odom`), no? Or did you not use that part of the other code's path handler?

If we have them in the costmap's global frame (ie `odom`), then we can use `mapToWorld` and `worldToMap` easy checks to convert the path's coordinates into the costmap frame to do lookups. Or is this control law not being computed in `odom` frame? 

I think something here is being over thought. We should be doing all the work in the `odom`, local costmap frame

Reply:
> Edit: oh, is it because the control law is in robot base coordinate frame? That would make sense

Yeap, you're right. This get the transform of the robot in the map to pass it to the trajectory simulation. That is because the control law assumes that all coordinates are in the robot base frame, egocentric, as they called in the paper.

Reply:
Got it!